compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingInterface.cpp

```diff
@@ -0,0 +1,537 @@
+      makeComposedAffineApply(b, loc, stepMap, ValueRange{step, nprocs});
+  return {distributeLB, distributeStep};
+}
+
+//===----------------------------------------------------------------------===//
+// TileDispatchUsingSCFForOp pattern implementation.
+//===----------------------------------------------------------------------===//
+
+namespace {
+// Till `scf.for` can take an `OpFoldResult` for lb, ub and step, use of `Range`
+// with `OpFoldResult` causes lots of `OpFoldResult` -> `Value` conversion.
+// This struct is a mirror of `Range` with `Value` type fields.
+struct RangeVal {
+  Value offset;
+  Value size;
+  Value stride;
+};
+}  // namespace
+
+/// Generate an empty loop nest that represents the tiled loop nest shell.
+/// - `loopRanges` specifies the lb, ub and step of the untiled iteration space.
+/// - `tileSizeVals` is the tile sizes to use. Zero represent untiled loops.
+/// - In `offsets` and `sizes` return the multi-dimensional offset and size of
+/// the
+///   tile processed within the inner most loop.
+static SmallVector<scf::ForOp> generateTileLoopNest(
+    OpBuilder &builder, Location loc, ArrayRef<RangeVal> loopRanges,
+    ArrayRef<Value> tileSizeVals,
+    ArrayRef<linalg::DistributionMethod> distributionMethod,
+    ArrayRef<linalg::ProcInfo> procInfo, SmallVector<OpFoldResult> &offsets,
+    SmallVector<OpFoldResult> &sizes) {
+  assert(!loopRanges.empty() && "expected at least one loop range");
+  assert(loopRanges.size() == tileSizeVals.size() &&
+         "expected as many tile sizes as loop ranges");
+  assert(loopRanges.size() == distributionMethod.size() &&
+         "expected as many entries in distribution method list as number of "
+         "loop ranges");
+  OpBuilder::InsertionGuard guard(builder);
+  SmallVector<scf::ForOp> loops;
+  offsets.resize(loopRanges.size());
+  sizes.resize(loopRanges.size());
+
+  // The tile size to use (to avoid out of bounds access) is  minimum of
+  // `tileSize` and `ub - iv`, where `iv` is the induction variable
+  // of the tiled loop.
+  AffineExpr s0, s1, d0;
+  bindDims(builder.getContext(), d0);
+  bindSymbols(builder.getContext(), s0, s1);
+  AffineMap minMap = AffineMap::get(1, 2, {s0, s1 - d0}, builder.getContext());
+  auto createBoundedTileSize = [&](Value iv, Value tileSize,
+                                   Value size) -> OpFoldResult {
+    if (isConstantIntValue(getAsOpFoldResult(tileSize), 1)) {
+      return builder.getIndexAttr(1);
+    }
+    return makeComposedFoldedAffineMin(
+        builder, loc, minMap, ArrayRef<OpFoldResult>{iv, tileSize, size});
+  };
+
+  unsigned procDim = 0;
+  for (auto loopRange : llvm::enumerate(loopRanges)) {
+    Value lb = loopRange.value().offset;
+    Value ub = loopRange.value().size;
+    Value step = tileSizeVals[loopRange.index()];
+
+    // No loops if tile size is zero. Set offset and size to the loop
+    // offset and size.
+    if (matchPattern(tileSizeVals[loopRange.index()], m_Zero())) {
```

Comment:
[optional] `loopRange.index()` is used many times, maybe worth to declare a variable for it?